### PR TITLE
Implement HDFS gateway

### LINKED ISSUE #142: Support HDFS Gateway
**Is your feature request related to a problem? Please describe.**

My company uses HDFS w/ Kerberos auth as a storage for artifacts and want to use it easily in argo-events.

**Describe the solution you'd like**

Create the HDFS gateway.

**Describe alternatives you've considered**

Create a custom gateway image by ourselves and specify it in `deploySpec`.

**Additional context**

I'm creating it for my company by myself, but I don't know if I should create a PR to the upstream or just create another repository for the gateway to share it publicly. HDFS is a quite common storage, but adding every common gateway implementation in the upstream may increase the effort of maintenance. So may I have your opinion?

### PR DESCRIPTION
**Pull request by dtaniwaki**
Resolve #142

I implemented the HDFS gateway.

I wasn't able to use [HDFS Inotify](http://hadoop.apache.org/docs/r3.2.0/api/org/apache/hadoop/hdfs/inotify/package-summary.html)  because it requires admin role and notifies everything happened in the cluster. Instead, I implemented naive file watcher which periodically scans the state of a watching directory and notify when the state is changed.

### CODE REVIEW

**Comment by dtaniwaki** [comment]
I'm rebasing the branch...

File: examples/gateways/hdfs.yaml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      labels:
+        gateway-name: "hdfs-gateway"
+    spec:
+      containers:
+      - name: "gateway-client"
+        image: "argoproj/gateway-client"
+        imagePullPolicy: "Always"
+        command: ["/bin/gateway-client"]
+      - name: "hdfs-events"
+        image: "argoproj/hdfs-gateway"
```
[the remark this answers is not in the document]
Comment (dtaniwaki) [reply examples/gateways/hdfs.yaml]: I’m totally fine with changing it, but would you tell me the reason? I think other community gateway images are hosted in argoproj.

[the remark this answers is not in the document]
Comment (VaibhavPage) [reply examples/gateways/hdfs.yaml]: Actually, there was a reshuffle in argoproj docker org and argo-events was added as a team. But that disabled members of argo-events team to create a new repo. I can ask Ed to create a repo for hdfs-gateway. :)

**Comment by dtaniwaki** [comment]
@VaibhavPage Thank you for reviewing the PR! I found a minor bug just now. Can I make a small change in this PR?

**Comment by dtaniwaki** [comment]
@VaibhavPage Updated, thanks!

**Pull request merged**